### PR TITLE
keyRel2

### PR DESCRIPTION
**Pull request by tom-wa**
Key relation types: Below, DirectBelow, Silbling, and Nephew
DirectBelow and Silbling are boolean tests
relation suffixes:
SN: the namespaces have to match
IN: ignoring namespaces completely
CN: cascading namespace matches all namespaces

currently returns: 
-1 for invalid keys / usage error
0 for failed tests
1 if a boolean relation test succeeded
>0 if non-boolean relation tests succeeded - the value is the number of levels the check key is below

### CODE REVIEW

File: src/include/kdb.h.in [hunk 1]
```diff
@@ -147,7 +147,28 @@ ssize_t    keySetMeta(Key *key, const char* metaName,
 
 /* Methods for Making Tests */
 int keyCmp(const Key *k1, const Key *k2);
+
+
+typedef enum
+{
+    BSN,        //Below Same Namespace, cascading namespace matches only cascading namespace
```
Comment (markus2330) [inline src/include/kdb.h.in]: should be in proposal first. And please suffix the names.

File: src/libs/elektra/keytest.c [hunk 2]
```diff
@@ -531,7 +531,113 @@ int keyRel (const Key * key, const Key * check)
 	return -2;
 }
 
+Key * keyAsCascading (const Key * key)
```
Comment (markus2330) [inline src/libs/elektra/keytest.c]: docu? What does the function do?

File: src/libs/elektra/keytest.c [hunk 3]
```diff
@@ -531,7 +531,113 @@ int keyRel (const Key * key, const Key * check)
...
+			}
+			else
+			{
+				return keyNew (ptr, KEY_CASCADING_NAME, KEY_END);
+			}
+		}
+	}
+}
+
+int keyGetLevelsBelow (const Key * key, const Key * check)
```
Comment (markus2330) [inline src/libs/elektra/keytest.c]: again

File: tests/abi/testabi_key.c [hunk 4]
```diff
@@ -2638,6 +2638,128 @@ static void test_keyCanonify ()
 	keyDel (k);
 }
 
+static void test_keyAsCascading ()
```
Comment (markus2330) [inline tests/abi/testabi_key.c]: A new file for tests is better, please in proposal tests.

File: src/include/kdb.h.in [hunk 5]
```diff
@@ -147,7 +147,28 @@ ssize_t    keySetMeta(Key *key, const char* metaName,
...
+    SILBSN,     //Silbling Same Namespace
+    SILBIN,     //Silbling Ignore Namespace
+    SILBCN,     //Silbling (allow) Cascading Namespace
+    NEPHSN,     //Nephew Same Namespace
+    NEPHIN,     //Nephew Ignore Namespace
+    NEPHCN,     //Nephew (allow) Cascading Namespace
+}KeyRelType;
+
 int keyRel (const Key *k1, const Key *k2);
+int keyRel2 (const Key *k1, const Key *k2, KeyRelType which);
```
Comment (markus2330) [inline src/include/kdb.h.in]: The idea is to replace `keyRel` with a new ABI version.

File: src/include/kdb.h.in [hunk 6]
```diff
@@ -147,7 +147,28 @@ ssize_t    keySetMeta(Key *key, const char* metaName,
...
+Key *keyAsCascading(const Key *key);
```
Comment (markus2330) [inline src/include/kdb.h.in]: I dont think that these two are useful enough for the main API.

Comment (tom-wa) [inline src/include/kdb.h.in]: you are right, i just needed them internally and test them for myself.

**Comment by markus2330** [comment]
Its always good if you write what you want to know directly in the PR description :-) Again, its hard to say what you wanted to achieve. Please put proper API docu in the source.

The description of the return values seem to be flawed: what if a test fails and the keys are not the same? And I need to get used to the enum values... Maybe a bit longer would be better.

File: src/libs/elektra/keytest.c [hunk 7]
```diff
@@ -531,7 +531,113 @@ int keyRel (const Key * key, const Key * check)
...
+	}
+	keyDel (toCheck);
+	return levels;
+}
 
+int keyRel2 (const Key * key, const Key * check, KeyRelType which)
+{
+	if (!key || !check) return -1;
+	if (!key->key || !check->key) return -1;
+	if (!keyCmp (key, check)) return 0;
```
Comment (markus2330) [inline src/libs/elektra/keytest.c]: I would avoid this, users can check with keyCmp themselves.

**Comment by tom-wa** [comment]
you are right about the return values. not even sure if checking if the keys are the same is needed here. 
and 0 for failed check and dropping the -1 would better anyway i guess.

**Comment by tom-wa** [comment]
where should i put the testcases for keyRel2 ?

**Comment by tom-wa** [comment]
nvm, just found test_proposal.c

i've updated it now, and used longer names for the enum values

File: tests/ctest/test_proposal.c [hunk 8]
```diff
@@ -64,5 +186,9 @@ int main (int argc, char ** argv)
 	test_ksPopAtCursor ();
 	test_ksToArray ();
 
+    test_keyAsCascading ();
```
Comment (markus2330) [inline tests/ctest/test_proposal.c]: please run code formatter

File: src/include/kdbproposal.h [hunk 9]
```diff
@@ -66,6 +66,30 @@ KeySet * elektraKeyGetMetaKeySet (const Key * key);
 Key * ksPrev (KeySet * ks);
 Key * ksPopAtCursor (KeySet * ks, cursor_t c);
 
+
+
+typedef enum
+{
+    BelowSameNS,        //Below Same Namespace, cascading namespace matches only cascading namespace
```
Comment (markus2330) [inline src/include/kdbproposal.h]: please use suffix and make them uppercase.

E.g. ELEKTRA_REL_BELOW_SAME_NS

**Comment by tom-wa** [comment]
@markus2330 done

**Comment by markus2330** [comment]
Thank you, well done!

**Pull request merged**